DataStax CDM 5.1.0 origin DB connection issue.

We found an issue with CDM `5.1.0` jar when trying to submit a spark job with origin DB as Cassasndra and target DB as astra _(please refer configuration parameters as below)_ over Dataproc Serverless batch we get following error.

Configuration parameters for Spark Job:
```
gcloud dataproc batches submit spark
--region=us-east1
--jars=gs://<GCP BUCKET PATH>/cassandra-data-migrator-5.1.0.jar
--class=com.datastax.cdm.job.Migrate
--properties
^#^spark.executor.memory=4g#
spark.executor.cores=4#
spark.executor.instances=2#
spark.cdm.perfops.batchSize=1#
spark.cdm.perfops.numParts=10#
spark.cdm.perfops.ratelimit.origin=15000#
spark.cdm.trackRun=true#
spark.cdm.trackRun.runId=2204#
spark.cdm.schema.origin.keyspaceTable=<SOURCE_KEYSPACE>.forecasting_features#
spark.cdm.connect.origin.username=<ORIGIN USERNAME>#
spark.cdm.connect.origin.password=<ORIGIN PASSWORD>#
spark.cdm.connect.origin.host=<ORIGIN HOST FQDN/ IP>#
spark.cdm.connect.origin.port=9042#
spark.cdm.schema.origin.column.writetime.names=feature#
spark.cdm.feature.constantColumns.names=tenant_uuid#
spark.cdm.feature.constantColumns.types=uuid#
spark.cdm.feature.constantColumns.values=e72a66e5-e185-4111-8f73-4c1d60b2d8a7#
spark.cdm.schema.target.keyspaceTable=<TARGET_KEYSPACE>.forecasting_features#
spark.cdm.connect.target.username=<TARGET USERNAME>#
spark.cdm.connect.target.password=<TARGET PASSWORD>#
spark.cdm.connect.target.scb=secure-connect-kpi-keng03-dev-diksha-astra-db-05004-mykronos.zip    
--files=gs://<GCP BUCKET PATH>/secure-connect-kpi-keng03-dev-diksha-astra-db-05004-mykronos.zip,gs://<GCP BUCKET PATH>/identity.jks,gs://<GCP BUCKET PATH>/trustStore.jks
--service-account=<GCP SECRET ACCOUNT NAME> 
--tags=<NETWORK_TAGS>
--subnet=<SUBNET>
--batch=batch-cdm-dse-22-11-008
--version=2.2
```

Error Logs:
```
Caused by: com.datastax.oss.driver.api.core.AllNodesFailedException: Could not reach any contact point, make sure you've provided valid addresses (showing first 3 nodes, use getAllErrors() for more): Node(endPoint=65f333dc-991b-4a9a-877c-21e250a2d67c-us-east1.db.int.dev.mykronos.com/<unresolved>:29042:00958fd0-af98-391f-9f51-7e3fe763da5c, hostId=null, hashCode=3e8d557f): [com.datastax.oss.driver.api.core.auth.AuthenticationException: Authentication error on node 65f333dc-991b-4a9a-877c-21e250a2d67c-us-east1.db.int.dev.mykronos.com/<unresolved>:29042:00958fd0-af98-391f-9f51-7e3fe763da5c: server replied with 'We recently improved your database security. To find out more and reconnect, see https://docs.datastax.com/en/astra/docs/manage-application-tokens.html' to AuthResponse request], Node(endPoint=65f333dc-991b-4a9a-877c-21e250a2d67c-us-east1.db.int.dev.mykronos.com/<unresolved>:29042:b4f6d9ed-0f1d-3f7a-82f1-a4a7ea4f84d4, hostId=null, hashCode=753592c3): [com.datastax.oss.driver.api.core.auth.AuthenticationException: Authentication error on node 65f333dc-991b-4a9a-877c-21e250a2d67c-us-east1.db.int.dev.mykronos.com/<unresolved>:29042:b4f6d9ed-0f1d-3f7a-82f1-a4a7ea4f84d4: server replied with 'We recently improved your database security. To find out more and reconnect, see https://docs.datastax.com/en/astra/docs/manage-application-tokens.html' to AuthResponse request], Node(endPoint=65f333dc-991b-4a9a-877c-21e250a2d67c-us-east1.db.int.dev.mykronos.com/<unresolved>:29042:678e6a1d-e5b1-3cba-b949-7bfaf823957f, hostId=null, hashCode=1ba02720): [com.datastax.oss.driver.api.core.auth.AuthenticationException: Authentication error on node 65f333dc-991b-4a9a-877c-21e250a2d67c-us-east1.db.int.dev.mykronos.com/<unresolved>:29042:678e6a1d-e5b1-3cba-b949-7bfaf823957f: server replied with 'We recently improved your database security. To find out more and reconnect, see https://docs.datastax.com/en/astra/docs/manage-application-tokens.html' to AuthResponse request]
    at com.datastax.oss.driver.api.core.AllNodesFailedException.copy(AllNodesFailedException.java:143)
    at com.datastax.oss.driver.internal.core.util.concurrent.CompletableFutures.getUninterruptibly(CompletableFutures.java:151)
    at com.datastax.oss.driver.api.core.session.SessionBuilder.build(SessionBuilder.java:837)
    at com.datastax.spark.connector.cql.DefaultConnectionFactory$.createSession(CassandraConnectionFactory.scala:143)
    at com.datastax.spark.connector.cql.CassandraConnector$.createSession(CassandraConnector.scala:167)
    ... 25 more
    Suppressed: com.datastax.oss.driver.api.core.auth.AuthenticationException: Authentication error on node
```

--- a/RELEASE.md
+++ b/RELEASE.md
@@ -1,4 +1,7 @@
 # Release Notes
+## [5.1.3] - 2024-11-27
+- Bug fix: Fixed connection issue caused when using different types of origin and target clusters (e.g. Cassandra/DSE with host/port and Astra with SCB).
+
 ## [5.1.2] - 2024-11-26
 - Bug fix: SCB file on some Spark worker nodes may get deleted before the connection is established, which may cause connection exception on that worker node. Added a static async SCB delete delay to address such issues.
 

--- a/src/main/java/com/datastax/cdm/cql/statement/TargetUpsertRunDetailsStatement.java
+++ b/src/main/java/com/datastax/cdm/cql/statement/TargetUpsertRunDetailsStatement.java
@@ -66,10 +66,13 @@ public class TargetUpsertRunDetailsStatement {
         // TODO: Remove this code block after a few releases, its only added for backward compatibility
         try {
             this.session.execute("ALTER TABLE " + cdmKsTabInfo + " ADD status TEXT");
+        } catch (Exception e) { // ignore if column already exists
+            logger.debug("Column 'status' already exists in table {}", cdmKsTabInfo);
+        }
+        try {
             this.session.execute("ALTER TABLE " + cdmKsTabDetails + " ADD run_info TEXT");
-        } catch (Exception e) {
-            // ignore if column already exists
-            logger.trace("Column 'status' already exists in table {}", cdmKsTabInfo);
+        } catch (Exception e) { // ignore if column already exists
+            logger.debug("Column 'run_info' already exists in table {}", cdmKsTabDetails);
         }
 
         boundInitInfoStatement = bindStatement("INSERT INTO " + cdmKsTabInfo

--- a/src/main/scala/com/datastax/cdm/job/BaseJob.scala
+++ b/src/main/scala/com/datastax/cdm/job/BaseJob.scala
@@ -17,7 +17,7 @@ package com.datastax.cdm.job
 
 import com.datastax.cdm.properties.{KnownProperties, PropertyHelper}
 import com.datastax.spark.connector.cql.CassandraConnector
-import org.apache.spark.{SparkConf, SparkContext}
+import org.apache.spark.SparkContext
 import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.SparkSession
 import org.slf4j.LoggerFactory
@@ -40,7 +40,6 @@ abstract class BaseJob[T: ClassTag] extends App {
 
   var spark: SparkSession = _
   var sContext: SparkContext = _
-  var sc: SparkConf = _
   var propertyHelper: PropertyHelper = _
 
   var consistencyLevel: String = _
@@ -69,8 +68,7 @@ abstract class BaseJob[T: ClassTag] extends App {
       .appName(jobName)
       .getOrCreate()
     sContext = spark.sparkContext
-    sc = sContext.getConf
-    propertyHelper = PropertyHelper.getInstance(sc);
+    propertyHelper = PropertyHelper.getInstance(sContext.getConf);
 
     runId = propertyHelper.getLong(KnownProperties.RUN_ID)
     prevRunId = propertyHelper.getLong(KnownProperties.PREV_RUN_ID)
@@ -79,9 +77,9 @@ abstract class BaseJob[T: ClassTag] extends App {
       runId = System.nanoTime();
     }
     consistencyLevel = propertyHelper.getString(KnownProperties.READ_CL)
-    connectionFetcher = new ConnectionFetcher(sc, propertyHelper)
-    originConnection = connectionFetcher.getConnection(Side.ORIGIN, consistencyLevel, runId)
-    targetConnection = connectionFetcher.getConnection(Side.TARGET, consistencyLevel, runId)
+    connectionFetcher = new ConnectionFetcher(propertyHelper)
+    originConnection = connectionFetcher.getConnection(sContext.getConf, Side.ORIGIN, consistencyLevel, runId)
+    targetConnection = connectionFetcher.getConnection(sContext.getConf, Side.TARGET, consistencyLevel, runId)
 
     val hasRandomPartitioner: Boolean = {
       val partitionerName = originConnection.withSessionDo(_.getMetadata.getTokenMap.get().getPartitionerName)

--- a/src/main/scala/com/datastax/cdm/job/ConnectionFetcher.scala
+++ b/src/main/scala/com/datastax/cdm/job/ConnectionFetcher.scala
@@ -23,7 +23,7 @@ import com.datastax.cdm.data.DataUtility.generateSCB
 import com.datastax.cdm.data.PKFactory.Side
 
 // TODO: CDM-31 - add localDC configuration support
-class ConnectionFetcher(config: SparkConf, propertyHelper: IPropertyHelper) extends Serializable {
+class ConnectionFetcher(propertyHelper: IPropertyHelper) extends Serializable {
   val logger: Logger = LoggerFactory.getLogger(this.getClass.getName)
 
   def getConnectionDetails(side: Side): ConnectionDetails = {
@@ -63,7 +63,7 @@ class ConnectionFetcher(config: SparkConf, propertyHelper: IPropertyHelper) exte
     }
   }
 
-  def getConnection(side: Side, consistencyLevel: String, runId: Long): CassandraConnector = {
+  def getConnection(config: SparkConf, side: Side, consistencyLevel: String, runId: Long): CassandraConnector = {
     val connectionDetails = getConnectionDetails(side)
 
     logger.info("PARAM --  SSL Enabled: "+connectionDetails.sslEnabled);

--- a/src/main/scala/com/datastax/cdm/job/DiffData.scala
+++ b/src/main/scala/com/datastax/cdm/job/DiffData.scala
@@ -17,7 +17,7 @@ package com.datastax.cdm.job
 
 import com.datastax.cdm.feature.TrackRun
 import com.datastax.cdm.data.PKFactory.Side
-import com.datastax.cdm.properties.{KnownProperties, PropertyHelper}
+import com.datastax.cdm.properties.KnownProperties
 import com.datastax.cdm.job.IJobSessionFactory.JobType
 
 object DiffData extends BasePartitionJob {
@@ -34,6 +34,8 @@ object DiffData extends BasePartitionJob {
       var ma = new CDMMetricsAccumulator(jobType)
       sContext.register(ma, "CDMMetricsAccumulator")
       
+      val bcOriginConfig = sContext.broadcast(sContext.getConf)
+      val bcTargetConfig = sContext.broadcast(sContext.getConf)      
       val bcConnectionFetcher = sContext.broadcast(connectionFetcher)
       val bcPropHelper = sContext.broadcast(propertyHelper)
       val bcJobFactory = sContext.broadcast(jobFactory)
@@ -42,8 +44,8 @@ object DiffData extends BasePartitionJob {
 
       slices.foreach(slice => {
         if (null == originConnection) {
-    		originConnection = bcConnectionFetcher.value.getConnection(Side.ORIGIN, bcPropHelper.value.getString(KnownProperties.READ_CL), bcRunId.value)
-    		targetConnection = bcConnectionFetcher.value.getConnection(Side.TARGET, bcPropHelper.value.getString(KnownProperties.READ_CL), bcRunId.value)
+    		originConnection = bcConnectionFetcher.value.getConnection(bcOriginConfig.value, Side.ORIGIN, bcPropHelper.value.getString(KnownProperties.READ_CL), bcRunId.value)
+    		targetConnection = bcConnectionFetcher.value.getConnection(bcTargetConfig.value, Side.TARGET, bcPropHelper.value.getString(KnownProperties.READ_CL), bcRunId.value)
             trackRunFeature = targetConnection.withSessionDo(targetSession => new TrackRun(targetSession, bcKeyspaceTableValue.value))
         }
         originConnection.withSessionDo(originSession =>

--- a/src/main/scala/com/datastax/cdm/job/GuardrailCheck.scala
+++ b/src/main/scala/com/datastax/cdm/job/GuardrailCheck.scala
@@ -16,7 +16,7 @@
 package com.datastax.cdm.job
 
 import com.datastax.cdm.data.PKFactory.Side
-import com.datastax.cdm.properties.{KnownProperties, PropertyHelper}
+import com.datastax.cdm.properties.KnownProperties
 import com.datastax.cdm.job.IJobSessionFactory.JobType
 
 object GuardrailCheck extends BasePartitionJob {
@@ -32,13 +32,14 @@ object GuardrailCheck extends BasePartitionJob {
       var ma = new CDMMetricsAccumulator(jobType)
       sContext.register(ma, "CDMMetricsAccumulator")
       
+      val bcOriginConfig = sContext.broadcast(sContext.getConf)
       val bcConnectionFetcher = sContext.broadcast(connectionFetcher)
       val bcPropHelper = sContext.broadcast(propertyHelper)
       val bcJobFactory = sContext.broadcast(jobFactory)
 
       slices.foreach(slice => {
         if (null == originConnection) {
-    		originConnection = bcConnectionFetcher.value.getConnection(Side.ORIGIN, bcPropHelper.value.getString(KnownProperties.READ_CL), 0)
+    		originConnection = bcConnectionFetcher.value.getConnection(bcOriginConfig.value, Side.ORIGIN, bcPropHelper.value.getString(KnownProperties.READ_CL), 0)
         }
         originConnection.withSessionDo(originSession =>
             bcJobFactory.value.getInstance(originSession, null, bcPropHelper.value)

--- a/src/main/scala/com/datastax/cdm/job/Migrate.scala
+++ b/src/main/scala/com/datastax/cdm/job/Migrate.scala
@@ -18,7 +18,7 @@ package com.datastax.cdm.job
 import com.datastax.cdm.feature.TrackRun
 import com.datastax.cdm.job.CDMMetricsAccumulator
 import com.datastax.cdm.data.PKFactory.Side
-import com.datastax.cdm.properties.{KnownProperties, PropertyHelper}
+import com.datastax.cdm.properties.KnownProperties
 import com.datastax.cdm.job.IJobSessionFactory.JobType
 
 object Migrate extends BasePartitionJob {
@@ -34,7 +34,9 @@ object Migrate extends BasePartitionJob {
           jobFactory.getInstance(originSession, targetSession, propertyHelper).initCdmRun(runId, prevRunId, parts, trackRunFeature, jobType)))
       var ma = new CDMMetricsAccumulator(jobType)
       sContext.register(ma, "CDMMetricsAccumulator")
-
+      
+      val bcOriginConfig = sContext.broadcast(sContext.getConf)
+      val bcTargetConfig = sContext.broadcast(sContext.getConf)
       val bcConnectionFetcher = sContext.broadcast(connectionFetcher)
       val bcPropHelper = sContext.broadcast(propertyHelper)
       val bcJobFactory = sContext.broadcast(jobFactory)
@@ -43,8 +45,8 @@ object Migrate extends BasePartitionJob {
 
       slices.foreach(slice => {
         if (null == originConnection) {
-    		originConnection = bcConnectionFetcher.value.getConnection(Side.ORIGIN, bcPropHelper.value.getString(KnownProperties.READ_CL), bcRunId.value)
-    		targetConnection = bcConnectionFetcher.value.getConnection(Side.TARGET, bcPropHelper.value.getString(KnownProperties.READ_CL), bcRunId.value)
+    		originConnection = bcConnectionFetcher.value.getConnection(bcOriginConfig.value, Side.ORIGIN, bcPropHelper.value.getString(KnownProperties.READ_CL), bcRunId.value)
+    		targetConnection = bcConnectionFetcher.value.getConnection(bcTargetConfig.value, Side.TARGET, bcPropHelper.value.getString(KnownProperties.READ_CL), bcRunId.value)
             trackRunFeature = targetConnection.withSessionDo(targetSession => new TrackRun(targetSession, bcKeyspaceTableValue.value))
         }
         originConnection.withSessionDo(originSession =>

--- a/src/test/java/com/datastax/cdm/job/ConnectionFetcherTest.java
+++ b/src/test/java/com/datastax/cdm/job/ConnectionFetcherTest.java
@@ -45,7 +45,7 @@ public class ConnectionFetcherTest extends CommonMocks {
         commonSetupWithoutDefaultClassVariables();
         MockitoAnnotations.openMocks(this);
 
-        cf = new ConnectionFetcher(conf, propertyHelper);
+        cf = new ConnectionFetcher(propertyHelper);
     }
 
     @Test

